### PR TITLE
Google workspace timeout issue

### PR DESCRIPTION
**Pull request by v-sabiraj**
Required items, please complete
   
   Change(s):
   - See guidance below

   Reason for Change(s):
   - See guidance below

   Version Updated:
   - Required only for Detections/Analytic Rule templates
   - See guidance below

   Testing Completed:
  
image

### CODE REVIEW

**Comment by v-mchatla** [comment]
Hi
Still working on further changes.
thanks

File: Solutions/GoogleWorkspaceReports/Data Connectors/GWorkspaceReportsAPISentinelConnector/__init__.py [hunk 1]
```diff
@@ -249,55 +267,78 @@ def gen_chunks_with_latesttime(data,log_type):
             logging.error("Something wrong. Exception error text: {}".format(err))
     return latest_timestamp
 
+def process_result(result_obj, start_time, postactivity_list, line):
```
Comment (v-muuppugund) [inline Solutions/GoogleWorkspaceReports/Data Connectors/GWorkspaceReportsAPISentinelConnector/__init__.py:270]: Please add documentation for the code changes done in this file

File: Solutions/GoogleWorkspaceReports/Data Connectors/GWorkspaceReportsAPISentinelConnector/__init__.py [hunk 2]
```diff
@@ -142,25 +143,42 @@ def GetDates(logType):
...
         result = results.get('items', [])
         result_activities.extend(result)
-    if result_activities == None or len(result_activities) == 0:
-        logging.info("Logs not founded for {} activity".format(activity))
-        logging.info("Activity - {}, processing {} events)".format(activity, len(result_activities)))
-    else:
-        logging.info("Activity - {}, processing {} events)".format(activity, len(result_activities)))
-        return result_activities
+        logging.info("Name of Activity: {}, Number of events: {})".format(activity, len(result_activities)))
+        if result_activities == None or len(result_activities) == 0:
```
Comment (v-muuppugund) [inline Solutions/GoogleWorkspaceReports/Data Connectors/GWorkspaceReportsAPISentinelConnector/__init__.py:155]: Please check the code ql issue

File: Solutions/GoogleWorkspaceReports/Data Connectors/GWorkspaceReportsAPISentinelConnector/__init__.py [hunk 3]
```diff
@@ -142,25 +143,42 @@ def GetDates(logType):
...
+    try:
+        result_activities = []
+        service = build('admin', 'reports_v1', credentials=creds, cache_discovery=False)
+        results = service.activities().list(userKey='all', applicationName=activity,
+                                                maxResults=1000, startTime=start_time, endTime=end_time, pageToken=next_page_token).execute()
+        next_page_token = results.get('nextPageToken', None)
+        result = results.get('items', [])
+        result_activities.extend(result)
+        logging.info("Number of events {}".format(len(result_activities)))
+        if result_activities == None or len(result_activities) == 0:
```
Comment (v-muuppugund) [inline Solutions/GoogleWorkspaceReports/Data Connectors/GWorkspaceReportsAPISentinelConnector/__init__.py:174]: Please check code ql issue

File: Solutions/GoogleWorkspaceReports/Data Connectors/GWorkspaceReportsAPISentinelConnector/__init__.py [hunk 4]
```diff
@@ -142,25 +143,42 @@ def GetDates(logType):
     return json.loads(activity_list) if (isBlank(logType)) else (json.loads(activity_list)[logType],end_time)
 
 def get_result(activity,start_time, end_time):
-    result_activities = []
-    service = build('admin', 'reports_v1', credentials=creds, cache_discovery=False)
-    results = service.activities().list(userKey='all', applicationName=activity,
-                                              maxResults=1000, startTime=start_time, endTime=end_time).execute()
-    next_page_token = results.get('nextPageToken', None)
```
Comment (v-muuppugund) [inline Solutions/GoogleWorkspaceReports/Data Connectors/GWorkspaceReportsAPISentinelConnector/__init__.py:149]: Please add screen shots for the changes tested  in the customer environment

**Review by v-muuppugund — COMMENTED** [review]
Please attach test cases executed if any for all the scenarios

File: Solutions/GoogleWorkspaceReports/Data Connectors/GWorkspaceReportsAPISentinelConnector/__init__.py (hunk 1 above)
Comment (v-muuppugund) [reply Solutions/GoogleWorkspaceReports/Data Connectors/GWorkspaceReportsAPISentinelConnector/__init__.py:270]: Feedback Please add documentation at all the places where not present

Comment (v-mchatla) [reply Solutions/GoogleWorkspaceReports/Data Connectors/GWorkspaceReportsAPISentinelConnector/__init__.py:274]: updated

File: Solutions/GoogleWorkspaceReports/Data Connectors/GWorkspaceReportsAPISentinelConnector/__init__.py (hunk 2 above)
Comment (v-mchatla) [reply Solutions/GoogleWorkspaceReports/Data Connectors/GWorkspaceReportsAPISentinelConnector/__init__.py]: Fixed

File: Solutions/GoogleWorkspaceReports/Data Connectors/GWorkspaceReportsAPISentinelConnector/__init__.py [hunk 5]
```diff
@@ -249,55 +267,78 @@ def gen_chunks_with_latesttime(data,log_type):
...
+            if(isBlank(latest_timestamp)):
+                latest_timestamp = start_time
+                logging.info("The latest timestamp is same as the original start time {} - {}".format(line,latest_timestamp))
+                # Fetch the latest timestamp
+                logging.info("The latest timestamp got from api activity is {} - {}".format(line,latest_timestamp))
+        postactivity_list[line] = latest_timestamp
+        state = StateManager(connection_string)
+        state.post(str(json.dumps(postactivity_list)))
+    return latest_timestamp 
+
```
[the remark this answers is not in the document]
Comment (v-mchatla) [reply Solutions/GoogleWorkspaceReports/Data Connectors/GWorkspaceReportsAPISentinelConnector/__init__.py:306]: Updated

File: Solutions/GoogleWorkspaceReports/Data Connectors/GWorkspaceReportsAPISentinelConnector/__init__.py (hunk 3 above)
Comment (v-mchatla) [reply Solutions/GoogleWorkspaceReports/Data Connectors/GWorkspaceReportsAPISentinelConnector/__init__.py]: Fixed

File: Solutions/GoogleWorkspaceReports/Data Connectors/GWorkspaceReportsAPISentinelConnector/__init__.py [hunk 6]
```diff
@@ -142,25 +143,44 @@
...
         result = results.get('items', [])
         result_activities.extend(result)
-    if result_activities == None or len(result_activities) == 0:
-        logging.info("Logs not founded for {} activity".format(activity))
-        logging.info("Activity - {}, processing {} events)".format(activity, len(result_activities)))
-    else:
-        logging.info("Activity - {}, processing {} events)".format(activity, len(result_activities)))
-        return result_activities
+        logging.info("Name of Activity: {}, Number of events: {})".format(activity, len(result_activities)))
+        if result_activities == None or len(result_activities) == 0:
```
[the remark this answers is not in the document]
Comment (v-mchatla) [reply Solutions/GoogleWorkspaceReports/Data Connectors/GWorkspaceReportsAPISentinelConnector/__init__.py]: Fixed

File: Solutions/GoogleWorkspaceReports/Data Connectors/GWorkspaceReportsAPISentinelConnector/__init__.py [hunk 7]
```diff
@@ -142,25 +143,44 @@
...
+    try:
+        result_activities = []
+        service = build('admin', 'reports_v1', credentials=creds, cache_discovery=False)
+        results = service.activities().list(userKey='all', applicationName=activity,
+                                                maxResults=1000, startTime=start_time, endTime=end_time, pageToken=next_page_token).execute()
+        next_page_token = results.get('nextPageToken', None)
+        result = results.get('items', [])
+        result_activities.extend(result)
+        logging.info("Number of events {}".format(len(result_activities)))
+        if result_activities == None or len(result_activities) == 0:
```
[the remark this answers is not in the document]
Comment (v-mchatla) [reply Solutions/GoogleWorkspaceReports/Data Connectors/GWorkspaceReportsAPISentinelConnector/__init__.py]: Fixed

**Comment by v-mchatla** [comment]
Hi @v-muuppugund 
Author has confirmed that the code is working as expected, after restarting the function app.
image

image

Thanks

**Comment by v-mchatla** [comment]
> Hi Still working on further changes. thanks

Done with all the changes and get it verified by customer.

**Comment by v-mchatla** [comment]
> Please attach test cases executed if any for all the scenarios

We have tested the scenarios but don't have test plan to attach.

**Review by v-muuppugund — APPROVED** [review]
Approving with following suggestions
1.Retry count can be moved to function app configuration.
2.Documentation can be added for the newly changes added

**Review by v-muuppugund — COMMENTED** [review]
Please merge from master

**Review by v-muuppugund — APPROVED** [review]
Approving with below suggestions
- Retry pattern count at function app configuration
- Documentation for the new changes added

**Review by v-atulyadav — APPROVED** [review]

**Pull request merged**